src/compiler/checker.ts

```diff
@@ -5480,7 +5480,7 @@ namespace ts {
 
         function getLiteralTypeForText(flags: TypeFlags, text: string) {
             const map = flags & TypeFlags.StringLiteral ? stringLiteralTypes : numericLiteralTypes;
-            return map[text] || (map[text] = createLiteralType(flags, text));
+            return getOrUpdate(map, text, () => createLiteralType(flags, text));

```

Comment:
This introduces a new closure and environment record for every call to `getLiteralTypeForText`. The original approach does not.